### ISSUE TITLE
[Optimization / CUDA]: Enable `cuda_streaming` for differentiability

### ISSUE THREAD

**Issue by QuentinWach** [QuentinWach/beamz#225]
`cuda_streamed` cannot accelerate optimization yet. BeamZ explicitly compiles differentiable simulations with the JAX backend and rejects native CUDA because no adjoint rule exists: `differentiable.py`.

Installing the native component or setting:

```bash
BEAMZ_EXECUTION_BACKEND=cuda_streamed
```

will therefore not change the 501 gradient evaluations. It can accelerate the final detached evaluation, but not topology optimization.

This is a JAX FFI constraint: foreign kernels are opaque to autodiff until an explicit VJP/adjoint implementation is supplied. [JAX FFI documentation](https://docs.jax.dev/en/latest/ffi.html)

## Enabling streamed differentiation properly
The best architecture is a whole-simulation custom VJP, not a transpose rule for every individual CUDA launch:

1. Let the existing streamed CUDA program perform the primal forward solve.
2. Save a bounded set of field and CPML checkpoints on the GPU.
3. Add a reverse-time CUDA adjoint program that:
   - replays forward segments between checkpoints;
   - propagates field and CPML cotangents backward;
   - injects monitor-output cotangents;
   - accumulates `∂loss/∂ε`.
4. Wrap the forward and backward FFI targets with `jax.custom_vjp`.
5. Keep BeamZ’s Gaussian filter, SSP projection, objective composition, and Optax optimizer in ordinary JAX. Only the expensive FDTD operator crosses the native boundary.
6. Expose `backend="cuda_streamed"` through differentiable preparation, with `auto` selecting it when the adjoint-capable extension is available.
